File: 361-splitBinaryTreeAsMaxProduct.py
from treeNode import TreeNode
import math

ans=-math.inf
s=set()

def dfsSet(root):
    if not root:
        return 0
    
    left=dfsSet(root.left)
    right=dfsSet(root.right)
    s.add(left)
    s.add(right)
    s.add(left+right+root.val)
    
    return left+right+root.val

def maxValue(root):
    global ans
    T=sum(root)
    dfsSet(root)
    for i in s:
        ans=max(ans,i*(T-i))
    return  ans

def sum(root):
    if not root:
        return 0
    
    return root.val+sum(root.left)+sum(root.right)

def dfs(root,T):
    if not root:
        return 0
    cur=root.val+dfs(root.left,T)+dfs(root.right,T)
    global ans
    ans=max(ans,T*(T-cur))
    return cur

def maxProduct(root):
    T=sum(root)
    dfs(root,T)
    return ans


root=TreeNode(1)
L1=TreeNode(2)
R1=TreeNode(3)
LL2=TreeNode(4)
LR2=TreeNode(5)

RL2=TreeNode(6)

root.left=L1
root.right=R1

L1.left=LL2
L1.right=LR2

R1.left=RL2

print(maxValue(root))

File: 361-splitBinaryTreeAsMaxProduct.cpp
#include"treeNode.h"
#include<iostream>
/*
Given a binary tree root, consider deleting an edge in the tree so that the tree becomes disjoint with two trees.
 Then, take the sum of each subtree and multiply the two numbers. 
 Return the largest such product we can get after deleting one edge.

Given the root of a binary tree, split the binary tree into two subtrees by removing 

one edge such that the product of the sums of the subtrees is maximized. 
Return the maximum product of the sums of the two subtrees.*/

int sumOfTree(TreeNode* root)
{
    if(!root)
        return 0;
    return root->m_val+sumOfTree(root->left)+sumOfTree(root->right);
}

int dfs(TreeNode* root,int sum,int &ans)
{
    if(!root)
        return 0;

    int cur=dfs(root->left,sum,ans)+dfs(root->right,sum,ans)+root->m_val;
    ans=std::max(cur*(sum-cur),ans);
    return cur;
}


int maxProductOfSubtrees(TreeNode* root)
{
    int ans=INT_MIN;
    int sum=sumOfTree(root);
    dfs(root,sum,ans);
    return ans;
}


int main()
{
    // root=TreeNode(1)
    TreeNode *root=new TreeNode(1);
    TreeNode *L1=new TreeNode(2);
    TreeNode *R1=new TreeNode(3);
    TreeNode *LL2=new TreeNode(4);
    TreeNode *LR2=new TreeNode(5);
    TreeNode *RL2=new TreeNode(6);

    root->left=L1;
    root->right=R1;
    
    L1->left=LL2;
    L1->right=LR2;
    R1->left=RL2;

    std::cout<<maxProductOfSubtrees(root);
    return 0;
}